include/pybind11/cast.h

```diff
@@ -242,13 +267,18 @@ PYBIND11_NOINLINE inline detail::type_info* get_type_info(PyTypeObject *type) {
     return bases.front();
 }
 
-PYBIND11_NOINLINE inline detail::type_info *get_type_info(const std::type_info &tp,
+/// Return the type info for a given C++ type; on lookup failure can either throw or return nullptr.
+PYBIND11_NOINLINE inline detail::type_info *get_type_info(const std::type_index &tp,
                                                           bool throw_if_missing = false) {
+    std::type_index type_idx(tp);
     auto &types = get_internals().registered_types_cpp;
-
-    auto it = types.find(std::type_index(tp));
+    auto it = types.find(type_idx);
     if (it != types.end())
         return (detail::type_info *) it->second;
+    auto &locals = registered_local_types_cpp();
```

Comment:
I'm reasonably sure this is an ODR violation. `registered_local_types_cpp()` is located inside an anonymous namespace inside a header which means there's going to be one unique function per translation unit, i.e.:
```
a.cpp -> pybind11::detail::<unique_name_a>::registered_local_types_cpp()
b.cpp -> pybind11::detail::<unique_name_b>::registered_local_types_cpp()
```
On the other hand, `get_type_info()` is not in an anonymous namespace, so per ODR we are promising that there is just one of those for all transaltion units. But because of the call highlighted above, the body of `get_type_info()` will be defined differently in different translalation units, i.e. one will call `<unique_name_a>::registered_local_types_cpp()`, another `<unique_name_b>::`, etc.

A possible solution could be to move the definition of `registered_local_types_cpp` into the `PYBIND11_MODULE` macro (without the anon namespace) which would ensure a single definition per module. Although, I'm not sure if GCC will consider the function-local `static` variable as unique or shared among different modules.